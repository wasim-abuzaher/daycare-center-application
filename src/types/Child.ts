export interface Child {
    fullName: string;
    checked_in: boolean;
}
