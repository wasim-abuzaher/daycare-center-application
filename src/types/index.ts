export * from './Child';
export * from './Classroom';
export * from './Center';
export * from './AppStackParamList';
