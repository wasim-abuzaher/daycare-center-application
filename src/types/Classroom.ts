import {Child} from './Child';

export interface Classroom {
    children: Array<Child> | null;
    id: string;
    name: string;
}
