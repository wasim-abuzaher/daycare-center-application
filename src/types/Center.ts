import {Classroom} from './Classroom';

export interface Center {
    allClassroomsAccessible: boolean;
    classrooms: Array<Classroom>;
    id: string;
    name: string;
}
