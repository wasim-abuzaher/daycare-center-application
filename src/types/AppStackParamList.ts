import {Center} from './Center';

export type AppStackParamList = {
    Center: undefined;
    Classroom: {
        center: Center;
        index: number;
    };
};
