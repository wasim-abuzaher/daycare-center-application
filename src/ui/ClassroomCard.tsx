import React from 'react';
import {Classroom} from '../types';
import {StyleSheet, Text, TouchableOpacity, View} from 'react-native';
import MaterialIcons from 'react-native-vector-icons/MaterialIcons';

interface Props {
    classroom: Classroom;
    index: number;
    onClassroomClick: (id: number) => void;
}

export function ClassroomCard({
    classroom,
    index,
    onClassroomClick,
}: Props): React.ReactElement {
    /**
     * Function that returns the number of children who are checked in or checked out of the classroom,
     * or 0 if classroom doesn't have children
     * @param arg one of 'in' or 'out'
     */
    const getStat = (arg: string): number => {
        if (!classroom.children) {
            return 0;
        }

        const totalChildren = classroom.children.length;
        const checkedInChildren = classroom.children.filter(
            c => c.checked_in
        ).length;
        return arg === 'in'
            ? checkedInChildren
            : totalChildren - checkedInChildren;
    };

    return (
        <TouchableOpacity onPress={() => onClassroomClick(index)}>
            <View style={styles.container}>
                <Text style={styles.title}>{classroom.name}</Text>
                <View style={styles.stats}>
                    <Text style={styles.description}>
                        Children:{' '}
                        {classroom.children ? classroom.children.length : 0}
                    </Text>
                    <View style={styles.stats}>
                        <MaterialIcons name={'how-to-reg'} size={20} />
                        <Text style={styles.statDescription}>
                            {getStat('in')}
                        </Text>
                        <MaterialIcons name={'highlight-off'} size={20} />
                        <Text style={styles.statDescription}>
                            {getStat('out')}
                        </Text>
                    </View>
                </View>
            </View>
        </TouchableOpacity>
    );
}

const styles = StyleSheet.create({
    container: {
        display: 'flex',
        height: 120,
        justifyContent: 'center',
        paddingHorizontal: 24,
        borderBottomColor: '#c1c1c1',
        borderBottomWidth: 1,
    },
    title: {
        fontSize: 24,
        fontWeight: '600',
    },
    stats: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        alignItems: 'baseline',
    },
    description: {
        marginTop: 8,
        fontSize: 18,
        fontWeight: '400',
    },
    statDescription: {
        marginHorizontal: 4,
        fontSize: 18,
        fontWeight: '400',
    },
});
