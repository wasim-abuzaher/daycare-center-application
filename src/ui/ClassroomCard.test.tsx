import React from 'react';
import renderer from 'react-test-renderer';
import {fireEvent, render} from '@testing-library/react-native';
import {ClassroomCard} from './ClassroomCard';
import {Classroom} from '../types';

describe('renders ClassroomCard correctly', () => {
    it('should match snapshot', () => {
        const mockClassroom: Classroom = {
            children: [{fullName: 'John Doe', checked_in: false}],
            id: '1',
            name: 'Mock Class',
        };

        const onClassroomClickMock = jest.fn();

        const card = renderer.create(
            <ClassroomCard
                classroom={mockClassroom}
                onClassroomClick={onClassroomClickMock}
                index={4}
            />
        );

        expect(card).toMatchSnapshot();
    });

    it('should call onClassroomClick with correct argument', () => {
        const mockClassroom: Classroom = {
            children: [{fullName: 'John Doe', checked_in: false}],
            id: '1',
            name: 'Mock Class',
        };

        const onClassroomClickMock = jest.fn();

        const {getByText} = render(
            <ClassroomCard
                classroom={mockClassroom}
                index={4}
                onClassroomClick={onClassroomClickMock}
            />
        );

        fireEvent.press(getByText('Mock Class'));
        expect(onClassroomClickMock).toBeCalledWith(4);
    });
});
