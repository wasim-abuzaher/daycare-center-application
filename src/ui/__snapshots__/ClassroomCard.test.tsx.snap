// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ClassroomCard correctly should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  nativeID="animatedComponent"
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderBottomColor": "#c1c1c1",
        "borderBottomWidth": 1,
        "display": "flex",
        "height": 120,
        "justifyContent": "center",
        "paddingHorizontal": 24,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 24,
          "fontWeight": "600",
        }
      }
    >
      Mock Class
    </Text>
    <View
      style={
        Object {
          "alignItems": "baseline",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 18,
            "fontWeight": "400",
            "marginTop": 8,
          }
        }
      >
        Children:
         
        1
      </Text>
      <View
        style={
          Object {
            "alignItems": "baseline",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "400",
              "marginHorizontal": 4,
            }
          }
        >
          0
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "400",
              "marginHorizontal": 4,
            }
          }
        >
          1
        </Text>
      </View>
    </View>
  </View>
</View>
`;
